C++ backend fails when put inherited object in another object type

See the code below.

### Example
```nim
type
  Foo {.inheritable.} = object
  Foo2 = object of Foo
  Bar = object
    x: Foo2

var b = Bar()
```

C backend - OK on both Mac and Windows.
C++ backend on Mac - OK with refc;  failed with ORC/ARC, error message:
```
> nim cpp --gc:orc sss.nim
Hint: used config file '/Users/tommo/.choosenim/toolchains/nim-#devel/config/nim.cfg' [Conf]
Hint: used config file '/Users/tommo/.choosenim/toolchains/nim-#devel/config/config.nims' [Conf]
....
/Volumes/prj/sss.nim(7, 5) Hint: 'b' is declared but not used [XDeclaredButNotUsed]
CC: stdlib_system.nim
CC: sss.nim
/Users/tommo/.cache/nim/sss_d/@msss.nim.cpp:64:81: error: no matching constructor for initialization of
      'tyObject_Foo2__9ahL77ntrZsIwkb3d79aEmnw'
N_LIB_PRIVATE tyObject_Bar__sn3WATLZsEVBvn7zdG4hyg b__p4W8w5cyf34l5eeq3NRwoQ = {{{(&NTIv2__9ahL77ntrZsIwkb3d79aEmnw_)}}}
                                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/tommo/.cache/nim/sss_d/@msss.nim.cpp:47:8: note: candidate constructor (the implicit copy constructor) not viable: cannot convert
      initializer list argument to 'const tyObject_Foo2__9ahL77ntrZsIwkb3d79aEmnw'
struct tyObject_Foo2__9ahL77ntrZsIwkb3d79aEmnw : public tyObject_Foo__0leyIMOM9a7DLpMIMHk9cOYQ {
       ^
/Users/tommo/.cache/nim/sss_d/@msss.nim.cpp:47:8: note: candidate constructor (the implicit move constructor) not viable: cannot convert
      initializer list argument to 'tyObject_Foo2__9ahL77ntrZsIwkb3d79aEmnw'
struct tyObject_Foo2__9ahL77ntrZsIwkb3d79aEmnw : public tyObject_Foo__0leyIMOM9a7DLpMIMHk9cOYQ {
       ^
/Users/tommo/.cache/nim/sss_d/@msss.nim.cpp:47:8: note: candidate constructor (the implicit default constructor) not viable: requires 0
      arguments, but 1 was provided
1 error generated.
Error: execution of an external compiler program 'clang++ -c -std=gnu++14 -funsigned-char    -I'/Users/tommo/.choosenim/toolchains/nim-#devel/lib' -I/Volumes/prj -o /Users/tommo/.cache/nim/sss_d/@msss.nim.cpp.o /Users/tommo/.cache/nim/sss_d/@msss.nim.cpp' failed with exit code: 1

```
 
----
C++ backend on Windows - Failed, error message:
```
Hint: used config file 'C:\Users\tommo\.choosenim\toolchains\nim-#devel\config\nim.cfg' [Conf]
Hint: used config file 'C:\Users\tommo\.choosenim\toolchains\nim-#devel\config\config.nims' [Conf]
....
C:\prj\bug.nim(7, 5) Hint: 'b' is declared but not used [XDeclaredButNotUsed]
CC: bug.nim
C:\Users\tommo\nimcache\bug_d\@mbug.nim.cpp:84:121: error: could not convert '{{(& NTI__B6RJJsnBCpiv5jn359b7JyA_)}}' from '<brace-enclosed initializer list>' to 'tyObject_Foo2__B6RJJsnBCpiv5jn359b7JyA'
 N_LIB_PRIVATE tyObject_Bar__9byueIqvoUoAILn9aJtrZh1w b__ybC7cit9bewkcaDkLobKB9ag = {{{(&NTI__B6RJJsnBCpiv5jn359b7JyA_)}}}
                                                                                                                         ^
Error: execution of an external compiler program 'g++.exe -c -std=gnu++14 -funsigned-char  -w -fmax-errors=3 -fpermissive -mno-ms-bitfields   -IC:\Users\tommo\.choosenim\toolchains\nim-#devel\lib -IC:\prj -o C:\Users\tommo\nimcache\bug_d\@mbug.nim.cpp.o C:\Users\tommo\nimcache\bug_d\@mbug.nim.cpp' failed with exit code: 1
```

Tried with both devel and 1.4.4.

--- a/compiler/cgen.nim
+++ b/compiler/cgen.nim
@@ -376,8 +376,7 @@ proc genObjectInit(p: BProc, section: TCProcSection, t: PType, a: var TLoc,
     else:
       linefmt(p, section, "$1.m_type = $2;$n", [r, genTypeInfoV1(p.module, t, a.lode.info)])
   of frEmbedded:
-    # inheritance in C++ does not allow struct initialization: bug #18410
-    if not p.module.compileToCpp and optTinyRtti in p.config.globalOptions:
+    if optTinyRtti in p.config.globalOptions:
       var tmp: TLoc
       if mode == constructRefObj:
         let objType = t.skipTypes(abstractInst+{tyRef})

--- a/compiler/extccomp.nim
+++ b/compiler/extccomp.nim
@@ -89,7 +89,7 @@ compiler gcc:
     asmStmtFrmt: "__asm__($1);$n",
     structStmtFmt: "$1 $3 $2 ", # struct|union [packed] $name
     produceAsm: gnuAsmListing,
-    cppXsupport: "-std=gnu++14 -funsigned-char",
+    cppXsupport: "-std=gnu++17 -funsigned-char",
     props: {hasSwitchRange, hasComputedGoto, hasCpp, hasGcGuard, hasGnuAsm,
             hasAttribute})
 
@@ -116,7 +116,7 @@ compiler nintendoSwitchGCC:
     asmStmtFrmt: "asm($1);$n",
     structStmtFmt: "$1 $3 $2 ", # struct|union [packed] $name
     produceAsm: gnuAsmListing,
-    cppXsupport: "-std=gnu++14 -funsigned-char",
+    cppXsupport: "-std=gnu++17 -funsigned-char",
     props: {hasSwitchRange, hasComputedGoto, hasCpp, hasGcGuard, hasGnuAsm,
             hasAttribute})
 

--- a/tests/cpp/torc.nim
+++ b/tests/cpp/torc.nim
@@ -13,3 +13,13 @@ type
 proc p(): Option[O] = none(O)
 
 doAssert $p() == "none(O)"
+
+# bug #17351
+type
+  Foo = object of RootObj
+  Foo2 = object of Foo
+  Bar = object
+    x: Foo2
+
+var b = Bar()
+discard b

